proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,329 @@
+
+### React Native CLI support
+
+Ideally the CLI (or any other build tool) should be able to consume the `reactNativeManifest` section and warn at build time if a configuration is invalid (either by having an invalid value or by having a mixture of invalid configurations).
+
+It's outside the scope of this RFC to delve into details of which tool will implement which logic and we defer this to a future discussion.
+
+### React Native Directory
+
+The React Native direcotry can easily consume the `reactNativeManifest` by:
+1. Query the NPM registry with [their API](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md) to obtain the repository information.
+2. Query the repository information to obtain the `packge.json` and read the `reactNativeManifest` section.
+
+## Future proof and extensibility
+
+We believe that the `reactNativeManifest` section should be extensible as we believe new capabilities and features can be added there.
+Specifically, we envision to evolve `reactNativeManifest` or `reactNativeManifest.capabilites` as follows:
+
+1. Each addition of a new capability [requires a new RFC](https://github.com/react-native-community/discussions-and-proposals#proposals) to be approved.
+2. Each RFC should define a new section in the `reactNativeManifest` with what's the intended use case.
+3. Each RFC should define what are the sensible defaults and should refrain from breaking changes.
+4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability.
+
+## Complete Example
+
+Here we present a full example of how a `reactNativeManifest` section could look like for a library and an app.
+
+For an app the section will look as follows:
+    
+```json
+{
+  "name": "my-awesome-app",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "version": "1.0",
+    "type": "app",
+    "capabilities": {
+      "hermes": {
+        "enabled": true
+      },
+      "newArch": {
+        "enabled": true
+      },
+      "codegenConfig": {
+        "name": "AppSpecs",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.facebook.fbreact.specs"
+        }
+      }
+    }
+  }
+}
+```
+
+For a library the section will look as follows:
+
+```json
+{
+  "name": "my-awesome-library",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "version": "1.0",
+    "type": "library",
+    "capabilities": {
+      "hermes": {
+        "enabled": true
+      },
+      "newArch": {
+        "enabled": true
+      },
+      "codegenConfig": {
+        "name": "LibrarySpecs",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.facebook.fbreact.specs"
+        }
+      }
+    }
+  }
+}
+```
+
+### JSON Schema
+
+The JSON Schema will be published as soon as we find an agreement on the key names and semantics.
```

Comment:
Possibly publishing a validation library instead?  Something similar to [npmcli/package-json](https://www.npmjs.com/package/@npmcli/package-json) which users could use to parse, validate and evaluate the contents correctly (without having to home brew solutions).

Reply:
It's up to the various tools. The easiest would be to use https://www.npmjs.com/package/jsonschema inside the CLI and validate the `package.json` against the schema we provide.

Reply:
I've added a clarification about it in 68c906f36cef9af41a12172bc1c944900fb64f3b